src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/BannedAssembliesAnalyzer.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    public sealed class BannedAssembliesAnalyzer : SymbolAnalyzerBase
+    {
+        private static HashSet<string> BannedAssemblies = new HashSet<string>()
+        {
+            "System.Text.Json",
+            "Newtonsoft.Json",
+            "System.Collections.Immutable"
+        };
+
+        private static readonly string BannedAssembliesMessageArgs = string.Join(", ", BannedAssemblies);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Descriptors.AZC0014);
+
+        public override SymbolKind[] SymbolKinds { get; } = new[]
+        {
+            SymbolKind.Method,
+            SymbolKind.Field,
+            SymbolKind.Property,
+            SymbolKind.Parameter,
+            SymbolKind.Event,
+            SymbolKind.NamedType
+        };
+
+        public override void Analyze(ISymbolAnalysisContext context)
+        {
+            void CheckType(ITypeSymbol type, ISymbol symbol)
+            {
+                if (type is INamedTypeSymbol namedTypeSymbol)
+                {
+                    if (BannedAssemblies.Contains(type.ContainingAssembly.Name))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0014, symbol.Locations.First(), BannedAssembliesMessageArgs), symbol);
```

Comment:
Wouldn't it report for private methods as well?

Reply:
There is a check to skip non-public APIs